### ISSUE TITLE
Silent bad codegen when optimizing expression

### ISSUE THREAD

**Issue by briansull** [dotnet/coreclr#11574]
Repro:

Compile Repro.cs below:

.\Tools\dotnetcli\dotnet.exe Tools\csc.exe /o+ /out:Program.exe Repro.cs /r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Private.CoreLib.dll /r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Console.dll
/r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Runtime.dll

Repro.cs

```
using System;
class Program
{
    static byte[] s_arr2;
    static byte[] s_arr3;

    static void Init()
    {
        s_arr2 = new byte[] { 0x11, 0x12, 0x13 };
        s_arr3 = new byte[] { 0x21, 0x22, 0x33 };
    }

    static void Main(string[] args)
    {
        Init();

        byte[] arr1 = new byte[] { 2 };
        byte[] arr2 = s_arr2;
        byte[] arr3 = s_arr3;

        int len = arr1.Length + arr2.Length + arr3.Length;
        int cur = 0;
        Console.WriteLine("1: cur = {0}", cur);
        cur += arr1.Length;
        Console.WriteLine("2: cur += {0}, now {1}", arr1.Length, cur);
        cur += arr2.Length;
        Console.WriteLine("3: cur += {0}, now {1}", arr2.Length, cur);
        cur += arr3.Length;
        Console.WriteLine("4: cur += {0}, now {1}", arr3.Length, cur);
        Console.WriteLine("5: len is {0}", len);
    }
}
```

Run the Repro:
.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe Repro.exe
outputs:
1: cur = 0
2: cur += 1, now 1
3: cur += 3, now 6
4: cur += 3, now 7
5: len is 7

Problem is "3: cur += 3, now 6"
The value of 6 is incorrect and should be 4

**Comment by pgavlin** [comment]
This is Yet Another Remorphing Issue:

```
Considering CSE #04 [def=100, use=100, cost= 9] CSE Expression:
N006 (  3,  3) CSE #03 (def)[000037] ---X--------             /--*  arrLen    int    <l:$443, c:$442>
N005 (  1,  1)              [000036] ------------             |  \--*  lclVar    ref    V03 loc2         u:3 <l:$2c4, c:$105>
N007 (  9,  9) CSE #04 (def)[000039] ---X--------             *  +         int    <l:$483, c:$482>
N003 (  1,  1)              [000003] ------------             |  /--*  const     int    1 $42
N004 (  5,  5) CSE #02 (def)[000035] ---X--------             \--*  +         int    <l:$481, c:$480>
N002 (  3,  3) CSE #01 (def)[000033] ---X--------                \--*  arrLen    int    <l:$441, c:$440>
N001 (  1,  1)              [000032] ------------                   \--*  lclVar    ref    V02 loc1         u:3 <l:$2c3, c:$103>

Conservative CSE Promotion (CSE never live at call) (300 < 400)
cseRefCnt=300, aggressiveRefCnt=500, moderateRefCnt=400
defCnt=100, useCnt=100, cost=9, size=9
def_cost=2, use_cost=2, extra_no_cost=14, extra_yes_cost=0
CSE cost savings check (914 >= 400) passes

Promoting CSE:

lvaGrabTemp returning 25 (V25 rat0) (a long lifetime temp) called for ValNumCSE.

CSE #04 def at [000039] replaced in BB01 with def of V25
New refCnts for V25: refCnt =  1, refCntWtd = 1   
New refCnts for V25: refCnt =  2, refCntWtd = 2   
New refCnts for V04: refCnt =  4, refCntWtd = 4   
New refCnts for V25: refCnt =  3, refCntWtd = 3   
New refCnts for V02: refCnt =  5, refCntWtd = 5   
New refCnts for V03: refCnt =  5, refCntWtd = 5   
New refCnts for V25: refCnt =  4, refCntWtd = 4   
optValnumCSE morphed tree:
N010 (  1,  1)              [000323] ------------                /--*  lclVar    int    V25 cse0          <l:$483, c:$482>
N011 ( 10, 10)              [000324] -A-X--------             /--*  comma     int    <l:$483, c:$482>
N006 (  1,  1)              [000003] ------------             |  |     /--*  const     int    1 $42
N007 (  9,  9)              [000039] ---X--------             |  |  /--*  +         int    <l:$483, c:$482>
N004 (  3,  3) CSE #03 (def)[000037] ---X--------             |  |  |  |  /--*  arrLen    int    <l:$443, c:$442>
N003 (  1,  1)              [000036] ------------             |  |  |  |  |  \--*  lclVar    ref    V03 loc2         u:3 <l:$2c4, c:$105>
N005 (  7,  7) CSE #02 (def)[000035] ---X--------             |  |  |  \--*  +         int    <l:$481, c:$480>
N002 (  3,  3) CSE #01 (def)[000033] ---X--------             |  |  |     \--*  arrLen    int    <l:$441, c:$440>
N001 (  1,  1)              [000032] ------------             |  |  |        \--*  lclVar    ref    V02 loc1         u:3 <l:$2c3, c:$103>
N009 (  9,  9)              [000322] -A-X----R---             |  \--*  =         int    $VN.Void
N008 (  1,  1)              [000321] D------N----             |     \--*  lclVar    int    V25 cse0          <l:$483, c:$482>
N013 ( 10, 10)              [000041] -A-X----R---             *  =         int    <l:$483, c:$482>
N012 (  1,  1)              [000040] D------N----             \--*  lclVar    int    V04 loc3         d:3 <l:$483, c:$482>
```

The reassociation of the add at node 35 has changed the value produced by that node, thus invalidating its value number.

**Comment by pgavlin** [comment]
cc @russellhadley

**Comment by russellhadley** [comment]
We've had a continued stream of these on x64 as well as a bump in the issues on x86 during bring up.  I think the argument that we were past these has been settled to the contrary.

**Comment by RussKeldorph** [comment]
@pgavlin If someone else is working on this, feel free to reassign.

**Comment by pgavlin** [comment]
Nah, I already have a fix

**Comment by pgavlin** [comment]
Closed with #11579.
